File: scripts/translate-ui.js
import fs from "fs/promises";
import path from "path";
import { fileURLToPath } from "url";
import { translate } from "@vitalets/google-translate-api";

const __dirname = path.dirname(fileURLToPath(import.meta.url));

// Langues cibles principales (format ISO-639-1)
// Exclut "fr" car c'est la source
const TARGET_LOCALES = [
  "en", "es", "de", "it", "pt", "nl", "ru", "zh-CN", "ja", "ko",
  "ar", "hi", "tr", "pl", "sv", "da", "fi", "no", "el", "he",
  "th", "vi", "id", "cs", "hu", "ro"
];

const MESSAGES_DIR = path.join(__dirname, "..", "messages");

async function translateObject(obj, targetLang) {
  const result = {};
  for (const [key, value] of Object.entries(obj)) {
    if (typeof value === "object" && value !== null) {
      result[key] = await translateObject(value, targetLang);
    } else if (typeof value === "string") {
      try {
        const { text } = await translate(value, { from: "fr", to: targetLang });
        result[key] = text;
        // Petit délai pour éviter le rate-limiting
        await new Promise((r) => setTimeout(r, 100));
      } catch (e) {
        console.error(`Erreur traduction "${value}" vers ${targetLang}:`, e.message);
        result[key] = value; // Fallback
      }
    } else {
      result[key] = value;
    }
  }
  return result;
}

async function main() {
  const sourcePath = path.join(MESSAGES_DIR, "fr.json");
  const sourceRaw = await fs.readFile(sourcePath, "utf-8");
  const sourceData = JSON.parse(sourceRaw);

  for (const locale of TARGET_LOCALES) {
    const targetPath = path.join(MESSAGES_DIR, `${locale}.json`);
    
    // Si le fichier existe déjà, on le zappe pour gagner du temps
    try {
      await fs.access(targetPath);
      console.log(`[${locale}] existe déjà.`);
      continue;
    } catch {}

    console.log(`Traduction de fr vers ${locale}...`);
    try {
      const translatedData = await translateObject(sourceData, locale);
      await fs.writeFile(targetPath, JSON.stringify(translatedData, null, 2));
      console.log(`✅ ${locale} terminé.`);
    } catch (e) {
      console.error(`❌ Échec pour ${locale} :`, e);
    }
  }
}

main();

File: scripts/translate_ui.py
import json
import os
import sys
import time

try:
    from deep_translator import GoogleTranslator
except ImportError:
    print("Veuillez installer deep_translator: pip install deep-translator")
    sys.exit(1)

# Chemin vers messages
MESSAGES_DIR = os.path.join(os.path.dirname(__file__), "..", "messages")
COUNTRIES_FILE = os.path.join(os.path.dirname(__file__), "..", "src", "data", "countries.json")

def load_json(path):
    with open(path, "r", encoding="utf-8") as f:
        return json.load(f)

def save_json(path, data):
    with open(path, "w", encoding="utf-8") as f:
        json.dump(data, f, ensure_ascii=False, indent=2)

def translate_dict(d, translator):
    """Traduit récursivement un dictionnaire de chaînes."""
    res = {}
    for k, v in d.items():
        if isinstance(v, dict):
            res[k] = translate_dict(v, translator)
        elif isinstance(v, str):
            try:
                # Éviter les variables type {name}
                # Pour faire simple avec Google Translate, on traduit tel quel, mais ça peut casser {var}.
                # L'application actuelle n'utilise presque pas d'interpolation complexe, ou on les fixera à la main.
                translated = translator.translate(v)
                res[k] = translated
                time.sleep(0.1) # anti-rate-limit
            except Exception as e:
                print(f"Erreur de traduction pour '{v}': {e}")
                res[k] = v
        else:
            res[k] = v
    return res

def main():
    countries = load_json(COUNTRIES_FILE)
    unique_locales = list(set(c["locale"] for c in countries))
    print(f"Locales cibles : {unique_locales}")

    source_path = os.path.join(MESSAGES_DIR, "fr.json")
    if not os.path.exists(source_path):
        print(f"Fichier source introuvable: {source_path}")
        return

    source_data = load_json(source_path)

    for loc in unique_locales:
        target_path = os.path.join(MESSAGES_DIR, f"{loc}.json")
        if os.path.exists(target_path):
            print(f"[{loc}] existe déjà, on ignore.")
            continue
        
        print(f"Traduction de fr vers {loc}...")
        # Google Translate API utilise des codes similaires, mais parfois différent.
        # deep-translator accepte la plupart des codes ISO.
        try:
            translator = GoogleTranslator(source="fr", target=loc)
            translated_data = translate_dict(source_data, translator)
            save_json(target_path, translated_data)
            print(f"✅ {loc} traduit avec succès.")
        except Exception as e:
            print(f"❌ Erreur pour {loc}: {e}")
            # Si erreur, on copie la source par défaut pour éviter un crash serveur
            save_json(target_path, source_data)

if __name__ == "__main__":
    main()
